### PR TITLE
Query: Include: Collection SQL and generated code improvements

### PR DESCRIPTION
**Pull request by anpete**
- When generating collection include query models, the cloned parent subquery should explicitly project the required columns for join comparison and ordering.
- Also tweaked generated code to no longer capture QueryContext in closures.

### CODE REVIEW

**Comment by anpete** [comment]
Ping @smitpatel @maumar

File: test/EFCore.SqlServer.FunctionalTests/InheritanceSqlServerTest.cs [hunk 1]
```diff
@@ -26,15 +30,15 @@ public virtual void Common_property_shares_column()
                 var cokeType = context.Model.FindEntityType(typeof(Coke));
                 var teaType = context.Model.FindEntityType(typeof(Tea));
 
-                Assert.Equal("SugarGrams", cokeType.FindProperty("SugarGrams").Relational().ColumnName);
-                Assert.Equal("CaffeineGrams", cokeType.FindProperty("CaffeineGrams").Relational().ColumnName);
-                Assert.Equal("CokeCO2", cokeType.FindProperty("Carbination").Relational().ColumnName);
+                AssertSql("SugarGrams", cokeType.FindProperty("SugarGrams").Relational().ColumnName);
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/InheritanceSqlServerTest.cs]: this should be Assert.Equal still

File: test/EFCore.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs [hunk 2]
```diff
@@ -33,17 +33,7 @@ var customers
...
-    TRACKED: True
-(QueryContext queryContext) => IEnumerable<Customer> _ShapedQuery(
-    queryContext: queryContext, 
-    shaperCommandContext: SelectExpression: 
-        SELECT [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
-        FROM [Customers] AS [c]
-    , 
-    shaper: UnbufferedEntityShaper<Customer>
-)",
+'from Customer <generated>_0 in DbSet<Customer>",
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: we should either verify contents of optimized query model here, or remove the test altogether - in this form it doesn't provide any value over ```Queryable_simple``` test

File: test/EFCore.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs [hunk 3]
```diff
@@ -99,67 +89,8 @@ var customers
...
-            Void queryContext.QueryBuffer.StartTracking(
-                entity: entity, 
-                entityType: EntityType: Customer
-            )
-            return queryContext.QueryBuffer.IncludeCollection(0, Navigation: Customer.Orders (<Orders>k__BackingField, ICollection<Order>) Collection ToDependent Order Inverse: Customer 0 -1 1 -1 -1, Navigation: Order.Customer (<Customer>k__BackingField, Customer) ToPrincipal Customer Inverse: Orders 0 -1 2 -1 -1, EntityType: Order, value(Metadata.Internal.ClrICollectionAccessor`3[Specification.Tests.TestModels.Northwind.Customer,System.Collections.Generic.ICollection`1[Specification.Tests.TestModels.Northwind.Order],Specification.Tests.TestModels.Northwind.Order]), value(Metadata.Internal.ClrPropertySetter`2[Specification.Tests.TestModels.Northwind.Order,Specification.Tests.TestModels.Northwind.Customer]), True, entity, Convert(included[0]))
-        }
-    )
-)",
+'from Customer c in DbSet<Customer>"
+,
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: we should either verify contents of optimized query model here, or remove the test altogether - in this form it doesn't provide any value over Queryable_simple test

Comment (anpete) [inline test/EFCore.SqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: It is checking that we are logging everything up to an including the QueryModel, just no longer the specifics of the QM itself, which changes often.

File: test/EFCore.SqlServer.FunctionalTests/IncludeSqlServerTest.cs [hunk 4]
```diff
@@ -778,46 +796,48 @@ public override void Include_duplicate_collection(bool useString)
...
+    SELECT [c4].*
+    FROM [Customers] AS [c4]
+    ORDER BY [c4].[CustomerID]
+    OFFSET 2 ROWS FETCH NEXT 2 ROWS ONLY
+) AS [t4]
+ORDER BY [t3].[CustomerID]",
+                    Sql);
 
-@__p_0: 2
+                Assert.Contains(
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/IncludeSqlServerTest.cs]: do we know why these joins are now performed on the client?

File: src/EFCore/Query/Internal/IncludeCompiler.cs [hunk 5]
```diff
@@ -716,107 +804,77 @@ private static Expression CreateKeyAccessExpression(Expression target, IReadOnly
...
 
-                toQueryModel.BodyClauses.Add(outerOrderByClause);
+                    if (ordering.Expression is MemberExpression memberExpression
+                        && memberExpression.Expression
+                            is QuerySourceReferenceExpression querySourceReferenceExpression
+                        && querySourceReferenceExpression.ReferencedQuerySource == querySource)
+                    {
+                        orderingExpressionAsPropertyExpression
+                            = EntityQueryModelVisitor
+                                .CreatePropertyExpression(
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/IncludeCompiler.cs]: Can we avoid this? Creating PropertyExpression without property.

File: test/EFCore.SqlServer.FunctionalTests/IncludeSqlServerTest.cs (hunk 4 above)
Comment (anpete) [inline test/EFCore.SqlServer.FunctionalTests/IncludeSqlServerTest.cs]: #8054

File: src/EFCore/Query/Internal/AnonymousObject.cs [hunk 6]
```diff
@@ -40,22 +49,42 @@ public struct CompositeKey
...
+                        => current + ((current * 397) ^ (argument?.GetHashCode() ?? 0)));
+            }
+        }
 
         /// <summary>
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
-        public override int GetHashCode() => 0;
+        public object GetValue(int index) => _values[index];
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/AnonymousObject.cs]: Can we get away with having this method? Is there any other way we can correlate inner expressions of AnonymousObject from outside?

File: src/EFCore/Query/ExpressionVisitors/Internal/NavigationRewritingExpressionVisitor.cs [hunk 7]
```diff
@@ -1208,7 +1208,7 @@ private static bool IsCompositeKey([NotNull] Type type)
         {
```
Comment (maumar) [inline src/EFCore/Query/ExpressionVisitors/Internal/NavigationRewritingExpressionVisitor.cs]: consider changing helper method name to ```IsAnonymousObject``` as welll

File: src/EFCore/Query/Internal/AnonymousObject.cs [hunk 8]
```diff
@@ -40,22 +49,42 @@ public struct CompositeKey
...
-        public override bool Equals(object obj)
-            => _values.SequenceEqual(((CompositeKey)obj)._values);
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                return _values.Aggregate(
+                    0,
+                    (current, argument)
+                        => current + ((current * 397) ^ (argument?.GetHashCode() ?? 0)));
```
Comment (anpete) [inline src/EFCore/Query/Internal/AnonymousObject.cs]: Yes.

File: src/EFCore/Query/Internal/AnonymousObject.cs (hunk 6 above)
Comment (anpete) [inline src/EFCore/Query/Internal/AnonymousObject.cs]: No.

File: src/EFCore/Query/ExpressionVisitors/Internal/NavigationRewritingExpressionVisitor.cs (hunk 7 above)
Comment (anpete) [inline src/EFCore/Query/ExpressionVisitors/Internal/NavigationRewritingExpressionVisitor.cs]: Considered it, but I think in this case it always represents a composite key and so is more descriptive.

File: src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs [hunk 9]
```diff
@@ -601,6 +601,37 @@ var boundExpression
...
+                        = (int)((ConstantExpression)methodCallExpression.Arguments.Single()).Value;
+
+                    var tableExpression
+                        = selectExpression
+                            .GetTableForQuerySource(
+                                querySourceReferenceExpression.ReferencedQuerySource);
+
+                    if (tableExpression is SelectExpression)
+                    {
+                        return selectExpression.Projection[projectionIndex];
```
Comment (anpete) [inline src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs]: In the first case we are translating the join predicate and so are binding directly to the join inner sequence query. In the second case we are translating the order by after the join and so are binding "through" the join expression, which has been lifted. Let me know if there is a better way here.

File: src/EFCore/Query/EntityQueryModelVisitor.cs [hunk 10]
```diff
@@ -639,7 +651,7 @@ private class IncludeRemovingExpressionVisitor : RelinqExpressionVisitor
         {
             protected override Expression VisitMethodCall(MethodCallExpression node)
                 => IncludeCompiler.IsIncludeMethod(node)
-                    ? node.Arguments[0]
```
Comment (anpete) [inline src/EFCore/Query/EntityQueryModelVisitor.cs]: Pre-pended queryContext parameter - was previously accessed via closure.

File: test/EFCore.SqlServer.FunctionalTests/InheritanceSqlServerTest.cs [hunk 11]
```diff
@@ -405,5 +412,10 @@ SELECT COUNT(*)
 ";
 
         private static string Sql => TestSqlLoggerFactory.Sql.Replace(Environment.NewLine, FileLineEnding);
+
+        private void AssertSql(string expected, string actual)
+        {
+            TestHelpers.AssertBaseline(expected, actual, _testOutputHelper);
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/InheritanceSqlServerTest.cs]: Not specific to this PR but why do we have actual as arg here? we always pass `Sql` as actual.

Comment (anpete) [inline test/EFCore.SqlServer.FunctionalTests/InheritanceSqlServerTest.cs]: Oh, nice. We should do that.

**Comment by anpete** [comment]
New version up.

**Comment by smitpatel** [comment]
Build failure
-- Missing NotNull annotations --
Microsoft.EntityFrameworkCore.Query.Internal.AnonymousObject.IsGetValueExpression[querySourceReferenceExpression]
Expected: False
Actual:   True

**Comment by anpete** [comment]
@maumar Bug in ApiConsistencyTest :trollface: out parameter! I am fixing.

**Pull request merged**